lib/membrane_h264_plugin/parser.ex

```diff
@@ -1,626 +1,556 @@
-                are present) the DTS was always bigger than PTS. If that is not desired,
-                you can set `add_dts_offset: false`.
-                """
-              ]
-
-  @impl true
-  def handle_init(_ctx, opts) do
-    output_stream_structure =
-      case opts.output_stream_structure do
-        :avc3 -> {:avc3, @nalu_length_size}
-        :avc1 -> {:avc1, @nalu_length_size}
-        stream_structure -> stream_structure
-      end
+  @callback generate_stream_format(
+              parameter_sets(),
+              last_sent_stream_format :: stream_format(),
+              state()
+            ) :: stream_format() | nil
+
+  @doc """
+  Invoked for each parsed access unit.
+  """
+  @callback get_parameter_sets(AUSplitter.access_unit()) :: parameter_sets()
 
-    {au_timestamp_generator, framerate} =
-      if opts.generate_best_effort_timestamps do
-        {AUTimestampGenerator.new(opts.generate_best_effort_timestamps),
-         opts.generate_best_effort_timestamps.framerate}
-      else
-        {nil, nil}
-      end
+  @doc """
+  Invoked for each parsed access unit.
 
-    state = %{
-      nalu_splitter: nil,
-      nalu_parser: nil,
-      au_splitter: AUSplitter.new(),
-      au_timestamp_generator: au_timestamp_generator,
-      framerate: framerate,
-      mode: nil,
-      profile: nil,
-      previous_buffer_timestamps: nil,
-      output_alignment: opts.output_alignment,
-      frame_prefix: <<>>,
-      skip_until_keyframe: opts.skip_until_keyframe,
-      repeat_parameter_sets: opts.repeat_parameter_sets,
-      cached_spss: %{},
-      cached_ppss: %{},
-      initial_spss: opts.spss,
-      initial_ppss: opts.ppss,
-      input_stream_structure: nil,
-      output_stream_structure: output_stream_structure
-    }
-
-    {[], state}
-  end
+  Returns `true` if the access unit is a keyframe.
+  """
+  @callback keyframe?(AUSplitter.access_unit()) :: boolean()
 
-  @impl true
-  def handle_stream_format(:input, stream_format, ctx, state) do
-    {alignment, input_raw_stream_structure} =
-      case stream_format do
-        %RemoteStream{type: :bytestream} ->
-          {:bytestream, :annexb}
+  @optional_callbacks parse_raw_input_stream_structure: 1,
+                      remove_parameter_sets_from_stream?: 1,
+                      generate_stream_format: 3,
+                      get_parameter_sets: 1,
+                      keyframe?: 1
 
-        %H264{alignment: alignment, stream_structure: stream_structure} ->
-          {alignment, stream_structure}
-      end
+  defmacro __using__(options) do
+    au_splitter = Keyword.fetch!(options, :au_splitter)
+    nalu_parser = Keyword.fetch!(options, :nalu_parser)
+    au_timestamp_generator = Keyword.fetch!(options, :au_timestamp_generator)
+    metadata_key = Keyword.fetch!(options, :metadata_key)
 
-    is_first_received_stream_format = is_nil(ctx.pads.input.stream_format)
+    quote location: :keep do
```

Comment:
This `quote` block generates _a lot_ of code, which is an [anti-pattern](https://hexdocs.pm/elixir/1.16.0-rc.0/macro-anti-patterns.html#large-code-generation-by-macros). Instead of that, I'd change this to something like `CommonParser` and change `handle_init`, `handle_process`, `handle_stream_format` and `handle_end_of_stream` to functions that would be explicitly called by particular parsers. I think it shouldn't be _that_ big change, while it would make the parsers simpler to reason about.

Reply:
We can add those functions directly on the `Parser` module ?

Reply:
Yes, I think so. I'm not sure if they should have the exact same names and arguments.

Reply:
This seems to be harder to refactor since the callbacks are buried deep inside the functions and are called by all the handlers. For now I moved most of the logic to the `Parser` module, however `credo` still complains about long quote block. I'll continue thinking of a way to refactor the code to make it easy to move to a common `Parser`.